src/awkward/_connect/numba/layoutbuilder.py

```diff
@@ -0,0 +1,2602 @@
+        return numba.types.StringLiteral(self._parameters)
+
+    @property
+    def length(self):
+        return numba.types.int64
+
+
+@numba.extending.typeof_impl.register(LayoutBuilder)
+def typeof_LayoutBuilder(val, c):
+    if isinstance(val, BitMasked):
+        return BitMaskedType(
+            numba.from_dtype(val._mask.dtype),
+            val._content,
+            val._valid_when,
+            val._lsb_order,
+            val._parameters,
+        )
+
+    elif isinstance(val, ByteMasked):
+        return ByteMaskedType(
+            numba.from_dtype(val._mask.dtype),
+            val._content,
+            val._valid_when,
+            val._parameters,
+        )
+
+    elif isinstance(val, Empty):
+        return EmptyType(val._parameters)
+
+    elif isinstance(val, IndexedOption):
+        return IndexedOptionType(
+            numba.from_dtype(val._index.dtype), val._content, val._parameters
+        )
+
+    elif isinstance(val, ListOffset):
+        return ListOffsetType(
+            numba.from_dtype(val._offsets.dtype), val._content, val._parameters
+        )
+
+    elif isinstance(val, Numpy):
+        return NumpyType(numba.from_dtype(val._data.dtype), val._parameters)
+
+    elif isinstance(val, Record):
+        return RecordType(val._contents, val._fields, val._parameters)
+
+    elif isinstance(val, Regular):
+        return RegularType(val._content, val._size, val._parameters)
+
+    elif isinstance(val, Tuple):
+        return TupleType(val._contents, val._parameters)
+
+    elif isinstance(val, Union):
+        return UnionType(
+            numba.from_dtype(val._index.dtype), val._contents, val._parameters
+        )
+
+    elif isinstance(val, Unmasked):
+        return UnmaskedType(val._content, val._parameters)
+
+    else:
+        raise TypeError("unrecognized LayoutBuilder type")
+
+
+@numba.extending.overload(len)
+def LayoutBuilderType_len(builder):
+    if isinstance(
+        builder,
+        (
+            BitMaskedType,
+            ByteMaskedType,
+            EmptyType,
+            IndexedOptionType,
+            ListOffsetType,
+            NumpyType,
+            RecordType,
+            RegularType,
+            TupleType,
+            UnionType,
+            UnmaskedType,
+        ),
+    ):
+
+        def len_impl(builder):
+            return builder._length_get()
+
+        return len_impl
+
+
+########## Numpy ############################################################
+
+
+@final
+class Numpy(LayoutBuilder):
+    def __init__(self, dtype, *, parameters=None, initial=1024, resize=8.0):
+        self._data = GrowableBuffer(dtype=dtype, initial=initial, resize=resize)
+        self._parameters = parameters
+
+    @classmethod
+    def _from_buffer(cls, data):
+        out = cls.__new__(cls)
+        out._data = data
+        out._parameters = None
+        return out
+
+    def __repr__(self):
+        return f"ak.numba.lb.Numpy({self._data.dtype}, parameters={self._parameters})"
```

Comment:
This qualified class name is not needed here, in `LayoutBuilder.__repr__`, or following, in `LayoutBuilder.type`, but it is needed in the `NumpyType.name`, because Numba uses that string to determine type equality. Fully qualifying it there makes sense, but here we can have a shorter string, though it would have to be one that tells the world that it's a LayoutBuilder "`Numpy`" and not some other kind of "`Numpy`".

Perhaps

```
    def __repr__(self):
        return f"layoutbuilder.Numpy({self._data.dtype}, parameters={self._parameters})"
```

Reply:
Although there's a rule that `reprs` that look like they could be copy-pasted as code should be copy-pasteable as code, that rule has to assume some imports. Usually, the assumption is that you've imported the class name as an unqualified word, so that

```python
    def __repr__(self):
        return f"{type(self).__name__}({', '.join(self._constructor_args)})"
```

is a good `repr`.

But the `{__module__}` in this case is going to be very long: `awkward.numba.layoutbuilder`. It won't even abbreviate `awkward` to `ak`.

Using "`layoutbuilder.`" as a prefix was imagining the hypothetical world in which the user has already imported

```python
from awkward.numba import layoutbuilder
```

just as NumPy's array dtypes assume that the user has already imported

```python
from numpy import array, dtype
```